File: leetcode/leetcode_easy/symmetric_tree.py
# Optional Imports:
from typing import List, Optional

# import collections

"""
Leetcode 101: Symmetric Tree

Given the root of a binary tree, check whether it is a mirror of itself (i.e., symmetric around its center).
"""

import unittest


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        def _sub(left: Optional[TreeNode], right: Optional[TreeNode]) -> bool:
            if left is None or right is None:
                return left == right

            elif left.val == right.val:
                if left.right == right.left and left.left == right.right:
                    return _sub(left.left, right.right) and _sub(left.right and right.left)
                else:
                    return False
            else:
                return False

        if (not root) or (root.left and not root.right) or (root.right and not root.left):
            return False
        else:
            if root.left == root.right:
                return _sub(root.left, root.right)
                


if __name__ == "__main__":
    unittest.main()

File: leetcode/leetcode_easy/symmetric_tree.c
/**
 * Leetcode 101: Symmetric Tree
 * Given the root of a binary tree, check whether it is a mirror of itself 
 * (i.e., symmetric around its center).
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

/**
 * Definition for a binary tree node.
 */
struct TreeNode 
{
        int val;
        struct TreeNode *left;
        struct TreeNode *right;
};

bool dfs ( struct TreeNode* left, struct TreeNode* right )
{
        if (left == NULL && right == NULL)
                return true;

        else if ((left == NULL && right != NULL) || (left != NULL && right == NULL))
              return false;

        else if (left->val == right->val)
        {
                return dfs(left->left, right->right) && dfs( left->right, right->left);
        }
        else
        {
                return false;
        }

}


bool isSymmetric(struct TreeNode* root) 
{
        if (root == NULL)
                return true;

        else if (root->left == NULL && root->right != NULL || root->left != NULL && root->right == NULL)
                return false;

        else
                return dfs ( root->left, root->right);
    
}
